File: src/analysis/env_builder/name_def_types.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

module Ast = Flow_ast
open Enclosing_context
open Hint
open Reason
open Loc_collections

type scope_kind =
  | Ordinary (* function or module *)
  | Async (* async function *)
  | Generator (* generator function *)
  | AsyncGenerator (* async generator function *)
  | Module (* module scope *)
  | DeclareModule (* module scope *)
  | DeclareNamespace (* namespace scope *)
  | Global (* global scope *)
  | Ctor (* constructor function *)
  | ComponentBody (* component syntax *)
[@@deriving show]

type class_stack = ALoc.t list

type for_kind =
  | In
  | Of of { await: bool }

(* A map from location of tparam to name. *)
type tparams_map = string ALocMap.t

type hint_node =
  | AnnotationHint of tparams_map * (ALoc.t, ALoc.t) Ast.Type.annotation
  | ValueHint of (enclosing_context * (ALoc.t, ALoc.t) Ast.Expression.t)
  | ProvidersHint of ALoc.t Nel.t
  | WriteLocHint of Env_api.With_ALoc.def_loc_type * ALoc.t
  | StringLiteralType of string
  | ReactFragmentType
  | ReactNodeType
  | AnyErrorHint of Reason.t
  | ComposedArrayPatternHint of ALoc.t * array_element_pattern_hint list
  | ComposedObjectPatternHint of ALoc.t * object_prop_pattern_hint list

and array_element_pattern_hint =
  | ArrayElementPatternHint of hint_node
  | ArrayRestElementPatternHint of hint_node

and object_prop_pattern_hint =
  | ObjectPropPatternHint of string * ALoc.t * hint_node
  | ObjectSpreadPropPatternHint of hint_node

type ast_hints =
  ( hint_node,
    (ALoc.t, ALoc.t) Ast.Expression.CallTypeArgs.t option,
    (ALoc.t, ALoc.t) Ast.Expression.ArgList.t,
    (ALoc.t, ALoc.t) Ast.JSX.Opening.attribute list * (ALoc.t * (ALoc.t, ALoc.t) Ast.JSX.child list)
  )
  hint
  list

type function_synth_kind =
  | FunctionSynthesizable
  | MissingReturn of ALoc.t

type object_missing_annot =
  | FuncMissingAnnot of ALoc.t
  | OtherMissingAnnot of ALoc.t

type object_synth_kind =
  | ObjectSynthesizable of {
      (* A set of this write locations that can be resolved by resolving the object. *)
      this_write_locs: Env_api.EnvSet.t;
    }
  | MissingMemberAnnots of { locs: object_missing_annot Nel.t }
  | Unsynthesizable

type dro_annot =
  | Hook
  | Comp

type value = {
  hints: ast_hints;
  expr: (ALoc.t, ALoc.t) Ast.Expression.t;
  decl_kind: Ast.Variable.kind option;
}

type root =
  | Annotation of {
      tparams_map: tparams_map;
      optional: bool;
      has_default_expression: bool;
      react_deep_read_only: dro_annot option;
      param_loc: ALoc.t option;
      annot: (ALoc.t, ALoc.t) Ast.Type.annotation;
      concrete: root option;
    }
  | Value of value
  | FunctionValue of {
      hints: ast_hints;
      synthesizable_from_annotation: function_synth_kind;
      function_loc: ALoc.t;
      function_: (ALoc.t, ALoc.t) Ast.Function.t;
      statics: Env_api.EnvKey.t SMap.t;
      arrow: bool;
      tparams_map: tparams_map;
    }
  | ObjectValue of {
      synthesizable: object_synth_kind;
      obj_loc: ALoc.t;
      obj: (ALoc.t, ALoc.t) Ast.Expression.Object.t;
    }
  | EmptyArray of {
      array_providers: ALocSet.t;
      arr_loc: ALoc.t;
    }
  | Contextual of {
      reason: Reason.reason;
      hints: ast_hints;
      optional: bool;
      default_expression: (ALoc.t, ALoc.t) Ast.Expression.t option;
    }
  | CatchUnannotated
  | UnannotatedParameter of Reason.t
  | For of for_kind * (ALoc.t, ALoc.t) Ast.Expression.t

type binding =
  | Root of root
  | Hooklike of binding
  | Select of {
      selector: Selector.t;
      parent: ALoc.t * binding;
    }

type import =
  | Named of {
      kind: Ast.Statement.ImportDeclaration.import_kind option;
      remote: string;
      local: string;
    }
  | Namespace of string
  | Default of string

type generator_annot = {
  tparams_map: tparams_map;
  return_annot: (ALoc.t, ALoc.t) Ast.Type.annotation;
  async: bool;
}

type expression_def = {
  cond_context: enclosing_context;
  chain: bool;
  expr: (ALoc.t, ALoc.t) Ast.Expression.t;
  hints: ast_hints;
}

type def =
  | Binding of binding
  | ExpressionDef of expression_def
  | MemberAssign of {
      member_loc: ALoc.t;
      member: (ALoc.t, ALoc.t) Ast.Expression.Member.t;
      rhs: (ALoc.t, ALoc.t) Ast.Expression.t;
    }
  | OpAssign of {
      exp_loc: ALoc.t;
      lhs: (ALoc.t, ALoc.t) Ast.Pattern.t;
      op: Ast.Expression.Assignment.operator;
      rhs: (ALoc.t, ALoc.t) Ast.Expression.t;
    }
  | Update of {
      exp_loc: ALoc.t;
      op: Ast.Expression.Update.operator;
    }
  | Function of {
      hints: ast_hints;
      synthesizable_from_annotation: function_synth_kind;
      arrow: bool;
      has_this_def: bool;
      function_loc: ALoc.t;
      function_: (ALoc.t, ALoc.t) Ast.Function.t;
      statics: Env_api.EnvKey.t SMap.t;
      tparams_map: tparams_map;
    }
  | Component of {
      tparams_map: tparams_map;
      component_loc: ALoc.t;
      component: (ALoc.t, ALoc.t) Ast.Statement.ComponentDeclaration.t;
    }
  | Class of {
      class_: (ALoc.t, ALoc.t) Ast.Class.t;
      class_loc: ALoc.t;
      (* A set of this and super write locations that can be resolved by resolving the class. *)
      this_super_write_locs: Env_api.EnvSet.t;
    }
  | DeclaredClass of ALoc.t * (ALoc.t, ALoc.t) Ast.Statement.DeclareClass.t
  | DeclaredComponent of ALoc.t * (ALoc.t, ALoc.t) Ast.Statement.DeclareComponent.t
  | TypeAlias of ALoc.t * (ALoc.t, ALoc.t) Ast.Statement.TypeAlias.t
  | OpaqueType of ALoc.t * (ALoc.t, ALoc.t) Ast.Statement.OpaqueType.t
  | TypeParam of {
      tparams_map: tparams_map;
      kind: Flow_ast_mapper.type_params_context;
      tparam: (ALoc.t, ALoc.t) Ast.Type.TypeParam.t;
    }
  | Interface of ALoc.t * (ALoc.t, ALoc.t) Ast.Statement.Interface.t
  | Enum of ALoc.t * string * ALoc.t Ast.Statement.EnumDeclaration.body
  | Import of {
      import_kind: Ast.Statement.ImportDeclaration.import_kind;
      import: import;
      source: string;
      source_loc: ALoc.t;
    }
  | GeneratorNext of generator_annot option
  | DeclaredNamespace of ALoc.t * (ALoc.t, ALoc.t) Ast.Statement.DeclareNamespace.t
  | MissingThisAnnot

module Print = struct
  open Utils_js

  let string_of_root = function
    | Contextual _ -> "contextual"
    | EmptyArray _ -> "[]"
    | CatchUnannotated -> "unannotated catch param"
    | UnannotatedParameter r -> Reason.string_of_reason r
    | Annotation { annot = (loc, _); _ } -> spf "annot %s" (ALoc.debug_to_string loc)
    | Value { expr = (loc, _); _ } -> spf "val %s" (ALoc.debug_to_string loc)
    | FunctionValue { function_loc; _ } -> spf "function val %s" (ALoc.debug_to_string function_loc)
    | ObjectValue _ -> "object"
    | For (In, (loc, _)) -> spf "for in %s" (ALoc.debug_to_string loc)
    | For (Of _, (loc, _)) -> spf "for of %s" (ALoc.debug_to_string loc)

  let string_of_selector = Selector.to_string

  let rec string_of_binding = function
    | Root r -> string_of_root r
    | Hooklike binding -> spf "(%s)<as hooklike>" (string_of_binding binding)
    | Select { selector; parent = (_, binding); _ } ->
      spf "(%s)%s" (string_of_binding binding) (string_of_selector selector)

  let string_of_import_kind =
    let open Ast.Statement.ImportDeclaration in
    function
    | ImportTypeof -> "typeof "
    | ImportType -> "type "
    | ImportValue -> ""

  let string_of_import = function
    | Named { kind; remote; local = _ } ->
      spf "%s%s" (Base.Option.value_map ~f:string_of_import_kind ~default:"" kind) remote
    | Namespace _ -> "namespace"
    | Default _ -> "default"

  let rec on_hint = function
    | AnnotationHint _ -> "annot hint"
    | ValueHint _ -> "value hint"
    | ProvidersHint _ -> "providers hint"
    | WriteLocHint _ -> "write loc hint"
    | StringLiteralType s -> "string literal hint: " ^ s
    | ReactFragmentType -> "react fragment hint"
    | ReactNodeType -> "react node hint"
    | AnyErrorHint _ -> "any type hint"
    | ComposedArrayPatternHint (_, elements) ->
      spf
        "[%s]"
        (elements
        |> Base.List.map ~f:(function
               | ArrayElementPatternHint h -> spf "(%s)" (on_hint h)
               | ArrayRestElementPatternHint h -> spf "...(%s)" (on_hint h)
               )
        |> Base.String.concat ~sep:", "
        )
    | ComposedObjectPatternHint (_, props) ->
      spf
        "{%s}"
        (props
        |> Base.List.map ~f:(function
               | ObjectPropPatternHint (n, _, h) -> spf "%s: %s" n (on_hint h)
               | ObjectSpreadPropPatternHint h -> spf "...(%s)" (on_hint h)
               )
        |> Base.String.concat ~sep:", "
        )

  let string_of_source = function
    | Binding b -> string_of_binding b
    | ExpressionDef { expr = (expr_loc, _); hints; _ } ->
      spf "exp %s (hint = %s)" (ALoc.debug_to_string expr_loc) (string_of_hints ~on_hint hints)
    | Update _ -> "[in/de]crement"
    | MemberAssign _ -> "member_assign"
    | OpAssign _ -> "opassign"
    | Function { function_ = { Ast.Function.id; _ }; _ } ->
      spf
        "fun %s"
        (Base.Option.value_map
           ~f:(fun (_, { Ast.Identifier.name; _ }) -> name)
           ~default:"<anonymous>"
           id
        )
    | Component
        {
          component = { Ast.Statement.ComponentDeclaration.id = (_, { Ast.Identifier.name; _ }); _ };
          _;
        } ->
      spf "component %s" name
    | DeclaredClass (_, { Ast.Statement.DeclareClass.id = (_, { Ast.Identifier.name; _ }); _ }) ->
      spf "declared class %s" name
    | DeclaredComponent
        (_, { Ast.Statement.DeclareComponent.id = (_, { Ast.Identifier.name; _ }); _ }) ->
      spf "declared component %s" name
    | Class { class_ = { Ast.Class.id; _ }; class_loc = _; this_super_write_locs = _ } ->
      spf
        "class %s"
        (Base.Option.value_map
           ~f:(fun (_, { Ast.Identifier.name; _ }) -> name)
           ~default:"<anonymous>"
           id
        )
    | TypeAlias (_, { Ast.Statement.TypeAlias.right = (loc, _); _ }) ->
      spf "alias %s" (ALoc.debug_to_string loc)
    | OpaqueType (_, { Ast.Statement.OpaqueType.id = (loc, _); _ }) ->
      spf "opaque %s" (ALoc.debug_to_string loc)
    | TypeParam { tparam = (loc, _); _ } -> spf "tparam %s" (ALoc.debug_to_string loc)
    | Enum (loc, name, _) -> spf "enum %s %s" name (ALoc.debug_to_string loc)
    | Interface _ -> "interface"
    | GeneratorNext _ -> "next"
    | DeclaredNamespace
        ( _,
          {
            Ast.Statement.DeclareNamespace.id =
              Ast.Statement.DeclareNamespace.Local (loc, { Ast.Identifier.name; _ });
            _;
          }
        ) ->
      spf "declare namespace %s %s" name (ALoc.debug_to_string loc)
    | DeclaredNamespace
        (_, { Ast.Statement.DeclareNamespace.id = Ast.Statement.DeclareNamespace.Global _; _ }) ->
      "declare global"
    | Import { import_kind; source; import; source_loc = _ } ->
      spf "import %s%s from %s" (string_of_import_kind import_kind) (string_of_import import) source
    | MissingThisAnnot -> "this (missing)"
end

type env_entries_map = (def * scope_kind * class_stack * ALoc.t virtual_reason) Env_api.EnvMap.t

type hint_map = ast_hints ALocMap.t

File: rust_port/crates/flow_env_builder/src/name_def_types.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeMap;
use std::collections::BTreeSet;

use flow_aloc::ALoc;
use flow_aloc::ALocMap;
use flow_common::enclosing_context::EnclosingContext;
use flow_common::hint::Hint;
use flow_common::reason::Reason;
use flow_common::reason::VirtualReason;
use flow_data_structure_wrapper::ord_map::FlowOrdMap;
use flow_data_structure_wrapper::ord_set::FlowOrdSet;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;
use flow_data_structure_wrapper::vector::FlowVector;
use flow_parser::ast;
use flow_parser::ast_visitor::TypeParamsContext;
use vec1::Vec1;

use crate::env_api::DefLocType;
use crate::env_api::EnvKey;
use crate::env_api::EnvMap;
use crate::env_api::EnvSet;
use crate::selector::Selector;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ScopeKind {
    Ordinary,                 // function or module
    Async,                    // async function
    Generator,                // generator function
    AsyncGenerator,           // async generator function
    Module,                   // module scope
    DeclareModule,            // module scope
    DeclareNamespace,         // namespace scope
    Global,                   // global scope
    Ctor,                     // constructor function
    ComponentOrHookBody,      // component or hook syntax
    AsyncComponentOrHookBody, // async component or hook syntax
}

pub type ClassStack = FlowVector<ALoc>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ForKind {
    In,
    Of { await_: bool },
}

/// A map from location of tparam to name.
pub type TparamsMap = FlowOrdMap<ALoc, FlowSmolStr>;

#[derive(Debug, Clone)]
pub enum HintNode {
    AnnotationHint(TparamsMap, ast::types::Annotation<ALoc, ALoc>),
    ValueHint(EnclosingContext, ast::expression::Expression<ALoc, ALoc>),
    ProvidersHint(Vec1<ALoc>),
    WriteLocHint(DefLocType, ALoc),
    StringLiteralType(FlowSmolStr),
    ReactFragmentType,
    ReactNodeType,
    AnyErrorHint(Reason),
    ComposedArrayPatternHint(ALoc, Vec<ArrayElementPatternHint>),
    ComposedObjectPatternHint(ALoc, Vec<ObjectPropPatternHint>),
}

#[derive(Debug, Clone)]
pub enum ArrayElementPatternHint {
    ArrayElementPatternHint(HintNode),
    ArrayRestElementPatternHint(HintNode),
}

#[derive(Debug, Clone)]
pub enum ObjectPropPatternHint {
    ObjectPropPatternHint(FlowSmolStr, ALoc, HintNode),
    ObjectSpreadPropPatternHint(HintNode),
}

pub type AstHint = Hint<
    'static,
    HintNode,
    Option<ast::expression::CallTypeArgs<ALoc, ALoc>>,
    ast::expression::ArgList<ALoc, ALoc>,
    (
        std::sync::Arc<[ast::jsx::OpeningAttribute<ALoc, ALoc>]>,
        (ALoc, Vec<ast::jsx::Child<ALoc, ALoc>>),
    ),
>;

pub type AstHints = Vec<AstHint>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FunctionSynthKind {
    FunctionSynthesizable,
    MissingReturn(ALoc),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ObjectMissingAnnot {
    FuncMissingAnnot(ALoc),
    OtherMissingAnnot(ALoc),
}

#[derive(Debug, Clone)]
pub enum ObjectSynthKind {
    ObjectSynthesizable {
        /// A set of this write locations that can be resolved by resolving the object.
        this_write_locs: EnvSet<ALoc>,
    },
    MissingMemberAnnots {
        locs: Vec1<ObjectMissingAnnot>,
    },
    Unsynthesizable,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DroAnnot {
    Hook,
    Comp,
}

#[derive(Clone)]
pub struct Value {
    pub hints: AstHints,
    pub expr: ast::expression::Expression<ALoc, ALoc>,
    pub decl_kind: Option<ast::VariableKind>,
    pub as_const: bool,
}

#[derive(Clone)]
pub struct AnnotationData {
    pub tparams_map: TparamsMap,
    pub optional: bool,
    pub has_default_expression: bool,
    pub react_deep_read_only: Option<DroAnnot>,
    pub param_loc: Option<ALoc>,
    pub annot: (ALoc, ast::types::Type<ALoc, ALoc>),
    pub concrete: Option<Box<Root>>,
}

#[derive(Clone)]
pub struct MatchCaseRootData {
    pub case_match_root_loc: ALoc,
    pub root_pattern_loc: ALoc,
    pub prev_pattern_loc: Option<ALoc>,
}

#[derive(Clone)]
pub struct FunctionValueData {
    pub hints: AstHints,
    pub synthesizable_from_annotation: FunctionSynthKind,
    pub function_loc: ALoc,
    pub function_: ast::function::Function<ALoc, ALoc>,
    pub statics: BTreeMap<FlowSmolStr, EnvKey<ALoc>>,
    pub arrow: bool,
    pub tparams_map: TparamsMap,
}

#[derive(Clone)]
pub struct ObjectValueData {
    pub synthesizable: ObjectSynthKind,
    pub obj_loc: ALoc,
    pub obj: ast::expression::Object<ALoc, ALoc>,
}

#[derive(Clone)]
pub struct EmptyArrayData {
    pub array_providers: FlowOrdSet<ALoc>,
    pub arr_loc: ALoc,
}

#[derive(Clone)]
pub struct ContextualData {
    pub reason: Reason,
    pub hints: AstHints,
    pub optional: bool,
    pub default_expression: Option<(ALoc, ast::expression::Expression<ALoc, ALoc>)>,
}

#[derive(Clone)]
pub enum Root {
    Annotation(Box<AnnotationData>),
    Value(Box<Value>),
    MatchCaseRoot(Box<MatchCaseRootData>),
    FunctionValue(Box<FunctionValueData>),
    ObjectValue(Box<ObjectValueData>),
    EmptyArray(Box<EmptyArrayData>),
    Contextual(Box<ContextualData>),
    CatchUnannotated,
    DeclareVariableMissingAnnotationAndInit,
    UnannotatedParameter(Reason),
    For(Box<(ForKind, (ALoc, ast::expression::Expression<ALoc, ALoc>))>),
}

pub fn mk_value(
    hints: Option<AstHints>,
    decl_kind: Option<ast::VariableKind>,
    as_const: bool,
    expr: ast::expression::Expression<ALoc, ALoc>,
) -> Root {
    Root::Value(Box::new(Value {
        hints: hints.unwrap_or_default(),
        expr,
        decl_kind,
        as_const,
    }))
}

#[derive(Clone)]
pub enum Binding {
    Root(Root),
    Hooklike(Box<Binding>),
    Select {
        selector: Selector<ALoc, ALoc>,
        parent: (ALoc, Box<Binding>),
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Import {
    Named {
        kind: Option<ast::statement::ImportKind>,
        remote: FlowSmolStr,
        local: FlowSmolStr,
    },
    Namespace(FlowSmolStr),
    Default(FlowSmolStr),
}

#[derive(Debug, Clone)]
pub struct GeneratorAnnot {
    pub tparams_map: TparamsMap,
    pub return_annot: (ALoc, ast::types::Type<ALoc, ALoc>),
    pub async_: bool,
}

#[derive(Clone)]
pub struct ExpressionDef {
    pub cond_context: EnclosingContext,
    pub chain: bool,
    pub expr: ast::expression::Expression<ALoc, ALoc>,
    pub hints: AstHints,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ClassKind {
    Class,
    Record {
        defaulted_props: BTreeSet<FlowSmolStr>,
    },
}

#[derive(Clone)]
pub struct MatchCasePatternData {
    pub case_match_root_loc: ALoc,
    pub has_guard: bool,
    pub pattern: (ALoc, ast::match_pattern::MatchPattern<ALoc, ALoc>),
    pub prev_pattern_loc: Option<ALoc>,
}

#[derive(Clone)]
pub struct MemberAssignData {
    pub member_loc: ALoc,
    pub member: ast::expression::Member<ALoc, ALoc>,
    pub rhs: (ALoc, ast::expression::Expression<ALoc, ALoc>),
}

#[derive(Clone)]
pub struct OpAssignData {
    pub exp_loc: ALoc,
    pub lhs: (ALoc, ast::pattern::Pattern<ALoc, ALoc>),
    pub op: ast::expression::AssignmentOperator,
    pub rhs: (ALoc, ast::expression::Expression<ALoc, ALoc>),
    pub assertion: bool,
}

#[derive(Clone)]
pub struct FunctionDefData {
    pub hints: AstHints,
    pub synthesizable_from_annotation: FunctionSynthKind,
    pub arrow: bool,
    pub has_this_def: bool,
    pub function_loc: ALoc,
    pub function_: ast::function::Function<ALoc, ALoc>,
    pub statics: BTreeMap<FlowSmolStr, EnvKey<ALoc>>,
    pub namespace_types: BTreeMap<FlowSmolStr, EnvKey<ALoc>>,
    pub tparams_map: TparamsMap,
}

#[derive(Clone)]
pub struct DeclaredFunctionDefData {
    pub declarations: Vec<(ALoc, ast::statement::DeclareFunction<ALoc, ALoc>)>,
    pub statics: BTreeMap<FlowSmolStr, EnvKey<ALoc>>,
    pub namespace_types: BTreeMap<FlowSmolStr, EnvKey<ALoc>>,
}

#[derive(Clone)]
pub struct ComponentDefData {
    pub tparams_map: TparamsMap,
    pub component_loc: ALoc,
    pub component: ast::statement::ComponentDeclaration<ALoc, ALoc>,
}

#[derive(Clone)]
pub struct ClassDefData {
    pub class_: ast::class::Class<ALoc, ALoc>,
    pub class_loc: ALoc,
    /// A set of this and super write locations that can be resolved by resolving the class.
    pub this_super_write_locs: EnvSet<ALoc>,
    pub kind: ClassKind,
}

#[derive(Clone)]
pub struct RecordDefData {
    pub record: ast::statement::RecordDeclaration<ALoc, ALoc>,
    pub record_loc: ALoc,
    /// A set of this and super write locations that can be resolved by resolving the record.
    pub this_super_write_locs: EnvSet<ALoc>,
    pub defaulted_props: BTreeSet<FlowSmolStr>,
}

#[derive(Clone)]
pub struct TypeParamData {
    pub tparams_map: TparamsMap,
    pub kind: TypeParamsContext,
    pub tparam: (ALoc, ast::types::TypeParam<ALoc, ALoc>),
}

#[derive(Clone)]
pub struct ImportData {
    pub import_kind: ast::statement::ImportKind,
    pub import: Import,
    pub source: FlowSmolStr,
    pub source_loc: ALoc,
}

#[derive(Clone)]
pub enum Def {
    Binding(Box<Binding>),
    MatchCasePattern(Box<MatchCasePatternData>),
    ExpressionDef(Box<ExpressionDef>),
    MemberAssign(Box<MemberAssignData>),
    OpAssign(Box<OpAssignData>),
    Update {
        exp_loc: ALoc,
        op: ast::expression::UpdateOperator,
    },
    Function(Box<FunctionDefData>),
    DeclaredFunction(Box<DeclaredFunctionDefData>),
    Component(Box<ComponentDefData>),
    Class(Box<ClassDefData>),
    Record(Box<RecordDefData>),
    DeclaredClass(ALoc, ast::statement::DeclareClass<ALoc, ALoc>),
    DeclaredComponent(ALoc, ast::statement::DeclareComponent<ALoc, ALoc>),
    TypeAlias(ALoc, ast::statement::TypeAlias<ALoc, ALoc>),
    OpaqueType(ALoc, ast::statement::OpaqueType<ALoc, ALoc>),
    TypeParam(Box<TypeParamData>),
    Interface(ALoc, ast::statement::Interface<ALoc, ALoc>),
    Enum(
        Box<(
            ALoc,
            FlowSmolStr,
            ast::statement::enum_declaration::Body<ALoc>,
        )>,
    ),
    Import(Box<ImportData>),
    GeneratorNext(Box<Option<GeneratorAnnot>>),
    DeclaredNamespace(ALoc, ast::statement::DeclareNamespace<ALoc, ALoc>),
    MissingThisAnnot,
}

pub mod print {
    use std::fmt::Write;

    use flow_aloc::ALoc;
    use flow_parser::loc_sig::LocSig;

    use super::*;

    pub fn string_of_root(root: &Root) -> String {
        match root {
            Root::Contextual(_) => "contextual".to_string(),
            Root::EmptyArray(_) => "[]".to_string(),
            Root::CatchUnannotated => "unannotated catch param".to_string(),
            Root::DeclareVariableMissingAnnotationAndInit => {
                "declare var missing annotation or init".to_string()
            }
            Root::UnannotatedParameter(r) => format!("{:?}", r),
            Root::Annotation(data) => {
                let (loc, _) = &data.annot;
                format!("annot {}", loc.debug_to_string(false))
            }
            Root::Value(val) => {
                format!("val {}", val.expr.loc().debug_to_string(false))
            }
            Root::MatchCaseRoot(data) => {
                format!(
                    "match root for case at {}",
                    data.case_match_root_loc.debug_to_string(false)
                )
            }
            Root::FunctionValue(data) => {
                format!("function val {}", data.function_loc.debug_to_string(false))
            }
            Root::ObjectValue(_) => "object".to_string(),
            Root::For(data) => match &data.0 {
                ForKind::In => format!("for in {}", data.1.0.debug_to_string(false)),
                ForKind::Of { .. } => format!("for of {}", data.1.0.debug_to_string(false)),
            },
        }
    }

    pub fn string_of_selector(sel: &Selector<ALoc, ALoc>) -> String {
        sel.to_string()
    }

    pub fn string_of_binding(binding: &Binding) -> String {
        match binding {
            Binding::Root(r) => string_of_root(r),
            Binding::Hooklike(b) => format!("({})<as hooklike>", string_of_binding(b)),
            Binding::Select {
                selector,
                parent: (_, binding),
                ..
            } => {
                format!(
                    "({}){}",
                    string_of_binding(binding),
                    string_of_selector(selector)
                )
            }
        }
    }

    pub fn string_of_import_kind(kind: &ast::statement::ImportKind) -> &'static str {
        use ast::statement::ImportKind;
        match kind {
            ImportKind::ImportTypeof => "typeof ",
            ImportKind::ImportType => "type ",
            ImportKind::ImportValue => "",
        }
    }

    pub fn string_of_import(import: &Import) -> String {
        match import {
            Import::Named { kind, remote, .. } => {
                let k = kind.as_ref().map(string_of_import_kind).unwrap_or("");
                format!("{}{}", k, remote)
            }
            Import::Namespace(_) => "namespace".to_string(),
            Import::Default(_) => "default".to_string(),
        }
    }

    pub fn on_hint(hint: &HintNode) -> String {
        match hint {
            HintNode::AnnotationHint(_, _) => "annot hint".to_string(),
            HintNode::ValueHint(_, _) => "value hint".to_string(),
            HintNode::ProvidersHint(_) => "providers hint".to_string(),
            HintNode::WriteLocHint(_, _) => "write loc hint".to_string(),
            HintNode::StringLiteralType(s) => format!("string literal hint: {}", s),
            HintNode::ReactFragmentType => "react fragment hint".to_string(),
            HintNode::ReactNodeType => "react node hint".to_string(),
            HintNode::AnyErrorHint(_) => "any type hint".to_string(),
            HintNode::ComposedArrayPatternHint(_, elements) => {
                let mut result = String::from("[");
                for (i, elem) in elements.iter().enumerate() {
                    if i > 0 {
                        result.push_str(", ");
                    }
                    match elem {
                        ArrayElementPatternHint::ArrayElementPatternHint(h) => {
                            write!(result, "({})", on_hint(h)).unwrap();
                        }
                        ArrayElementPatternHint::ArrayRestElementPatternHint(h) => {
                            write!(result, "...({})", on_hint(h)).unwrap();
                        }
                    }
                }
                result.push(']');
                result
            }
            HintNode::ComposedObjectPatternHint(_, props) => {
                let mut result = String::from("{");
                for (i, prop) in props.iter().enumerate() {
                    if i > 0 {
                        result.push_str(", ");
                    }
                    match prop {
                        ObjectPropPatternHint::ObjectPropPatternHint(n, _, h) => {
                            write!(result, "{}: {}", n, on_hint(h)).unwrap();
                        }
                        ObjectPropPatternHint::ObjectSpreadPropPatternHint(h) => {
                            write!(result, "...({})", on_hint(h)).unwrap();
                        }
                    }
                }
                result.push('}');
                result
            }
        }
    }

    pub fn string_of_source(def: &Def) -> String {
        match def {
            Def::Binding(b) => string_of_binding(b),
            Def::MatchCasePattern(data) => {
                let (loc, _) = &data.pattern;
                format!("match pattern {}", loc.debug_to_string(false))
            }
            Def::ExpressionDef(data) => {
                format!(
                    "exp {} (hint = {})",
                    data.expr.loc().debug_to_string(false),
                    string_of_hints(&data.hints)
                )
            }
            Def::Update { .. } => "[in/de]crement".to_string(),
            Def::MemberAssign(_) => "member_assign".to_string(),
            Def::OpAssign(_) => "opassign".to_string(),
            Def::Function(data) => {
                let name = data
                    .function_
                    .id
                    .as_ref()
                    .map(|id| id.name.as_str())
                    .unwrap_or("<anonymous>");
                format!("fun {}", name)
            }
            Def::Component(data) => {
                format!("component {}", data.component.id.name)
            }
            Def::DeclaredClass(_, decl) => {
                format!("declared class {}", decl.id.name)
            }
            Def::DeclaredComponent(_, decl) => {
                format!("declared component {}", decl.id.name)
            }
            Def::Class(data) => {
                let kind_str = match &data.kind {
                    ClassKind::Class => "class",
                    ClassKind::Record { .. } => "record",
                };
                let name = data
                    .class_
                    .id
                    .as_ref()
                    .map(|id| id.name.as_str())
                    .unwrap_or("<anonymous>");
                format!("{} {}", kind_str, name)
            }
            Def::Record(data) => {
                format!("record {}", data.record.id.name)
            }
            Def::DeclaredFunction(data) => {
                let name = data
                    .declarations
                    .first()
                    .and_then(|(_, decl)| decl.id.as_ref())
                    .map(|id| id.name.as_str())
                    .unwrap_or("<anonymous>");
                format!("declare function {}", name)
            }
            Def::TypeAlias(_, alias) => {
                format!("alias {}", alias.right.loc().debug_to_string(false))
            }
            Def::OpaqueType(_, opaque) => {
                format!("opaque {}", opaque.id.loc.debug_to_string(false))
            }
            Def::TypeParam(data) => {
                let (loc, _) = &data.tparam;
                format!("tparam {}", loc.debug_to_string(false))
            }
            Def::Enum(data) => {
                let (loc, name, _) = &**data;
                format!("enum {} {}", name, loc.debug_to_string(false))
            }
            Def::Interface(_, _) => "interface".to_string(),
            Def::GeneratorNext(_) => "next".to_string(),
            Def::DeclaredNamespace(_, ns) => {
                use ast::statement::declare_namespace::Id;
                match &ns.id {
                    Id::Local(id) => {
                        format!(
                            "declare namespace {} {}",
                            id.name,
                            id.loc.debug_to_string(false)
                        )
                    }
                    Id::Global(_) => "declare global".to_string(),
                }
            }
            Def::Import(data) => {
                format!(
                    "import {}{} from {}",
                    string_of_import_kind(&data.import_kind),
                    string_of_import(&data.import),
                    data.source
                )
            }
            Def::MissingThisAnnot => "this (missing)".to_string(),
        }
    }

    fn string_of_hints(hints: &AstHints) -> String {
        use flow_common::hint::Hint;
        let strs: Vec<String> = hints
            .iter()
            .map(|hint| match hint {
                Hint::HintT(t, _) => format!("Hint_t ({})", on_hint(t)),
                Hint::HintDecomp(_, t, _) => format!("Hint_Decomp ({})", on_hint(t)),
                Hint::HintPlaceholder => "Hint_Placeholder".to_string(),
            })
            .collect();
        format!("[{}]", strs.join(", "))
    }
}

pub type EnvEntriesMap = EnvMap<ALoc, (Def, ScopeKind, ClassStack, VirtualReason<ALoc>)>;

pub type HintMap = ALocMap<AstHints>;
